File: python/timestep/eval/suite.py
"""Suite runner for evaluation harness."""

import dataclasses
import random
from pathlib import Path
from typing import Any, Dict, List

from ..core.episode import run_episode, EpisodeInfo
from ..core.tools import ToolFn, index_tool_calls
from ..core.types import AgentFn, Message
from .graders import Grader, aggregate_grades
from ..utils.jsonl import read_jsonl, write_jsonl
from ..utils.io import write_json, now
from ..utils.messages import ensure_task_id

JSON = Dict[str, Any]


def run_suite(
    tasks_path: Path,
    outdir: Path,
    agent: AgentFn,
    tools: Dict[str, ToolFn],
    graders: List[Grader],
    trials: int,
    seed: int,
    agent_timeout_s: int,
) -> None:
    """Run evaluation suite on tasks from JSONL file."""
    outdir.mkdir(parents=True, exist_ok=True)
    rng = random.Random(seed)

    run_meta = {
        "version": "eval_mvp_v1",
        "tasks_path": str(tasks_path),
        "trials": trials,
        "seed": seed,
        "started_at": now(),
        "graders": [g.__class__.__name__ for g in graders],
        "agent_timeout_s": agent_timeout_s,
        "tools_available": sorted(list(tools.keys())),
    }
    write_json(outdir / "run_meta.json", run_meta)

    results_rows: List[JSON] = []

    for task in read_jsonl(tasks_path):
        task_id = ensure_task_id(task)
        task_messages = task.get("messages")
        if not isinstance(task_messages, list):
            raise SystemExit(f"Task {task_id} missing 'messages' list.")

        tools_allowed = task.get("tools_allowed")  # optional allowlist
        limits = task.get("limits", {}) or {}

        for trial in range(1, trials + 1):
            trial_seed = rng.randrange(0, 2**31 - 1)

            trial_dir = outdir / "trials" / task_id / f"trial_{trial:02d}"
            trial_dir.mkdir(parents=True, exist_ok=True)

            # Attach trial metadata (kept out of messages)
            task_meta = dict(task)
            task_meta["_trial"] = trial

            # Run episode (core agent-environment loop)
            messages, info = run_episode(
                initial_messages=task_messages,
                agent=agent,
                tools=tools,
                tools_allowed=tools_allowed,
                limits=limits,
                task_meta=task_meta,
                seed=trial_seed,
            )

            # Build tool index
            tool_idx = index_tool_calls(messages)

            # Grade
            grade_rows = [g.grade(messages, tool_idx, task, info) for g in graders]
            agg = aggregate_grades(grade_rows)

            # Persist artifacts
            write_json(trial_dir / "transcript.json", messages)
            write_json(trial_dir / "tool_index.json", [dataclasses.asdict(r) for r in tool_idx])
            write_json(trial_dir / "grades.json", {"grades": grade_rows, "aggregate": agg})
            write_json(trial_dir / "info.json", dataclasses.asdict(info))

            # Row for results.jsonl
            results_rows.append({
                "task_id": task_id,
                "trial": trial,
                "seed": trial_seed,
                "terminated_reason": info.terminated_reason,
                "steps": info.steps,
                "tool_calls": info.tool_calls,
                "duration_s": info.duration_s,
                "input_tokens": info.input_tokens,
                "output_tokens": info.output_tokens,
                "total_tokens": info.total_tokens,
                "cost_usd": info.cost_usd,
                "passed": agg["passed"],
                "score": agg["score"],
            })

    write_jsonl(outdir / "results.jsonl", results_rows)
    run_meta["ended_at"] = now()
    write_json(outdir / "run_meta.json", run_meta)


def report(outdir: Path) -> None:
    """Print summary report of evaluation results."""
    results_path = outdir / "results.jsonl"
    if not results_path.exists():
        raise SystemExit(f"No results.jsonl in {outdir}")

    rows = list(read_jsonl(results_path))
    if not rows:
        print("No results.")
        return

    overall_pass = sum(1 for r in rows if r.get("passed")) / len(rows)
    overall_score = sum(float(r.get("score", 0.0)) for r in rows) / len(rows)
    avg_tokens = sum(float(r.get("total_tokens", 0)) for r in rows) / len(rows) if rows else 0

    by_task: Dict[str, List[JSON]] = {}
    for r in rows:
        by_task.setdefault(str(r["task_id"]), []).append(r)

    task_summaries = []
    for tid, rs in by_task.items():
        k = len(rs)
        passed_count = sum(1 for x in rs if x.get("passed"))
        pr = passed_count / k if k > 0 else 0.0
        ms = sum(float(x.get("score", 0.0)) for x in rs) / k if k > 0 else 0.0
        md = sum(float(x.get("duration_s", 0.0)) for x in rs) / k if k > 0 else 0.0
        mt = sum(float(x.get("total_tokens", 0)) for x in rs) / k if k > 0 else 0.0
        pass_at_k = 1.0 if passed_count > 0 else 0.0
        pass_power_k = 1.0 if passed_count == k else 0.0
        task_summaries.append((tid, pr, ms, md, mt, k, pass_at_k, pass_power_k))

    task_summaries.sort(key=lambda x: (x[1], x[2]))  # worst first

    # Overall pass@k and pass^k
    total_tasks = len(by_task)
    tasks_with_any_pass = sum(1 for _, rs in by_task.items() if any(r.get("passed") for r in rs))
    tasks_with_all_pass = sum(1 for _, rs in by_task.items() if all(r.get("passed") for r in rs))
    overall_pass_at_k = tasks_with_any_pass / total_tasks if total_tasks > 0 else 0.0
    overall_pass_power_k = tasks_with_all_pass / total_tasks if total_tasks > 0 else 0.0

    print(f"Run: {outdir}")
    print(f"Trials: {len(rows)}")
    print(f"Overall pass rate: {overall_pass:.3f}")
    print(f"Overall mean score: {overall_score:.3f}")
    print(f"Overall pass@k: {overall_pass_at_k:.3f}")
    print(f"Overall pass^k: {overall_pass_power_k:.3f}")
    print(f"Average tokens per trial: {avg_tokens:.0f}")
    print()
    
    # Format table with aligned columns
    header = f"{'task_id':<40} {'pass_rate':>10} {'mean_score':>11} {'pass@k':>8} {'pass^k':>8} {'duration_s':>11} {'tokens':>8} {'trials':>7}"
    print(f"Worst tasks:")
    print(f"  {header}")
    print(f"  {'-' * 40} {'-' * 10} {'-' * 11} {'-' * 8} {'-' * 8} {'-' * 11} {'-' * 8} {'-' * 7}")
    for tid, pr, ms, md, mt, k, pass_at_k, pass_power_k in task_summaries[:20]:
        # Truncate task_id if too long
        tid_display = tid[:37] + "..." if len(tid) > 40 else tid
        print(f"  {tid_display:<40} {pr:>10.3f} {ms:>11.3f} {pass_at_k:>8.3f} {pass_power_k:>8.3f} {md:>11.3f} {mt:>8.0f} {k:>7}")

File: typescript/timestep/eval/suite.ts
/** Suite runner for evaluation harness. */

import { readJsonl, writeJsonl } from '../utils/jsonl';
import { writeJson, now } from '../utils/io';
import { ensureTaskId } from '../utils/messages';
import type { AgentFn, ToolFn, EpisodeInfo, JSON } from '../core/index';
import { runEpisode, indexToolCalls } from '../core/index';
import type { Grader } from './graders';
import { aggregateGrades } from './graders';
import { mkdirSync } from 'fs';
import { join } from 'path';

export async function runSuite(
  tasksPath: string,
  outdir: string,
  agent: AgentFn,
  tools: Record<string, ToolFn>,
  graders: Grader[],
  trials: number,
  seed: number,
  agentTimeoutS: number,
): Promise<void> {
  /** Run evaluation suite on tasks from JSONL file. */
  mkdirSync(outdir, { recursive: true });
  
  // Simple seeded RNG
  let rngState = seed;
  function random(): number {
    rngState = (rngState * 1103515245 + 12345) & 0x7fffffff;
    return rngState / 0x7fffffff;
  }
  function randrange(min: number, max: number): number {
    return Math.floor(random() * (max - min)) + min;
  }

  const runMeta: JSON = {
    version: 'eval_mvp_v1',
    tasks_path: tasksPath,
    trials,
    seed,
    started_at: now(),
    graders: graders.map(g => g.name),
    agent_timeout_s: agentTimeoutS,
    tools_available: Object.keys(tools).sort(),
  };
  writeJson(join(outdir, 'run_meta.json'), runMeta);

  const resultsRows: JSON[] = [];

  for (const task of readJsonl(tasksPath)) {
    const taskId = ensureTaskId(task);
    const taskMessages = task.messages;
    if (!Array.isArray(taskMessages)) {
      throw new Error(`Task ${taskId} missing 'messages' list.`);
    }

    const toolsAllowed = task.tools_allowed; // optional allowlist
    const limits = task.limits || {};

    for (let trial = 1; trial <= trials; trial++) {
      const trialSeed = randrange(0, 2**31 - 1);

      const trialDir = join(outdir, 'trials', taskId, `trial_${String(trial).padStart(2, '0')}`);
      mkdirSync(trialDir, { recursive: true });

      // Attach trial metadata (kept out of messages)
      const taskMeta = { ...task, _trial: trial };

      // Run episode (core agent-environment loop)
      const [messages, info] = await runEpisode(
        taskMessages,
        agent,
        tools,
        toolsAllowed,
        limits,
        taskMeta,
        trialSeed,
      );

      // Build tool index
      const toolIdx = indexToolCalls(messages);

      // Grade (handle async graders)
      const gradeRows = await Promise.all(
        graders.map(async g => {
          const result = g.grade(messages, toolIdx, task, info);
          // Handle both sync and async graders
          return result instanceof Promise ? await result : result;
        })
      );
      const agg = aggregateGrades(gradeRows);

      // Persist artifacts
      writeJson(join(trialDir, 'transcript.json'), messages);
      writeJson(join(trialDir, 'tool_index.json'), toolIdx.map(r => ({
        tool_call_id: r.tool_call_id,
        name: r.name,
        arguments_raw: r.arguments_raw,
        arguments: r.arguments,
        result_raw: r.result_raw,
        result: r.result,
        error: r.error,
      })));
      writeJson(join(trialDir, 'grades.json'), { grades: gradeRows, aggregate: agg });
      writeJson(join(trialDir, 'info.json'), {
        task_id: info.task_id,
        trial: info.trial,
        seed: info.seed,
        steps: info.steps,
        tool_calls: info.tool_calls,
        duration_s: info.duration_s,
        terminated_reason: info.terminated_reason,
        error: info.error,
        input_tokens: info.input_tokens,
        output_tokens: info.output_tokens,
        total_tokens: info.total_tokens,
        cost_usd: info.cost_usd,
      });

      // Row for results.jsonl
      resultsRows.push({
        task_id: taskId,
        trial,
        seed: trialSeed,
        terminated_reason: info.terminated_reason,
        steps: info.steps,
        tool_calls: info.tool_calls,
        duration_s: info.duration_s,
        input_tokens: info.input_tokens,
        output_tokens: info.output_tokens,
        total_tokens: info.total_tokens,
        cost_usd: info.cost_usd,
        passed: agg.passed,
        score: agg.score,
      });
    }
  }

  writeJsonl(join(outdir, 'results.jsonl'), resultsRows);
  runMeta.ended_at = now();
  writeJson(join(outdir, 'run_meta.json'), runMeta);
}

export function report(outdir: string): void {
  /** Print summary report of evaluation results. */
  const resultsPath = join(outdir, 'results.jsonl');
  
  let rows: JSON[] = [];
  try {
    rows = Array.from(readJsonl(resultsPath));
  } catch (e: any) {
    throw new Error(`No results.jsonl in ${outdir}: ${e.message}`);
  }
  
  if (rows.length === 0) {
    console.log('No results.');
    return;
  }

  const overallPass = rows.filter(r => r.passed).length / rows.length;
  const overallScore = rows.reduce((sum, r) => sum + Number(r.score || 0), 0) / rows.length;
  const avgTokens = rows.reduce((sum, r) => sum + Number(r.total_tokens || 0), 0) / rows.length;

  const byTask: Record<string, JSON[]> = {};
  for (const r of rows) {
    const tid = String(r.task_id);
    if (!byTask[tid]) byTask[tid] = [];
    byTask[tid].push(r);
  }

  const taskSummaries: Array<[string, number, number, number, number, number, number, number]> = [];
  for (const [tid, rs] of Object.entries(byTask)) {
    const k = rs.length;
    const passedCount = rs.filter(x => x.passed).length;
    const pr = passedCount / k;
    const ms = rs.reduce((sum, x) => sum + Number(x.score || 0), 0) / k;
    const md = rs.reduce((sum, x) => sum + Number(x.duration_s || 0), 0) / k;
    const mt = rs.reduce((sum, x) => sum + Number(x.total_tokens || 0), 0) / k;
    const passAtK = passedCount > 0 ? 1.0 : 0.0;
    const passPowerK = passedCount === k ? 1.0 : 0.0;
    taskSummaries.push([tid, pr, ms, md, mt, k, passAtK, passPowerK]);
  }

  taskSummaries.sort((a, b) => a[1] - b[1] || a[2] - b[2]); // worst first

  // Overall pass@k and pass^k
  const totalTasks = Object.keys(byTask).length;
  const tasksWithAnyPass = Object.values(byTask).filter(rs => rs.some(r => r.passed)).length;
  const tasksWithAllPass = Object.values(byTask).filter(rs => rs.every(r => r.passed)).length;
  const overallPassAtK = totalTasks > 0 ? tasksWithAnyPass / totalTasks : 0.0;
  const overallPassPowerK = totalTasks > 0 ? tasksWithAllPass / totalTasks : 0.0;

  console.log(`Run: ${outdir}`);
  console.log(`Trials: ${rows.length}`);
  console.log(`Overall pass rate: ${overallPass.toFixed(3)}`);
  console.log(`Overall mean score: ${overallScore.toFixed(3)}`);
  console.log(`Overall pass@k: ${overallPassAtK.toFixed(3)}`);
  console.log(`Overall pass^k: ${overallPassPowerK.toFixed(3)}`);
  console.log(`Average tokens per trial: ${avgTokens.toFixed(0)}`);
  console.log();
  
  // Format table with aligned columns
  const header = `${'task_id'.padEnd(40)} ${'pass_rate'.padStart(10)} ${'mean_score'.padStart(11)} ${'pass@k'.padStart(8)} ${'pass^k'.padStart(8)} ${'duration_s'.padStart(11)} ${'tokens'.padStart(8)} ${'trials'.padStart(7)}`;
  console.log('Worst tasks:');
  console.log(`  ${header}`);
  console.log(`  ${'-'.repeat(40)} ${'-'.repeat(10)} ${'-'.repeat(11)} ${'-'.repeat(8)} ${'-'.repeat(8)} ${'-'.repeat(11)} ${'-'.repeat(8)} ${'-'.repeat(7)}`);
  for (const [tid, pr, ms, md, mt, k, passAtK, passPowerK] of taskSummaries.slice(0, 20)) {
    // Truncate task_id if too long
    const tidDisplay = tid.length > 40 ? tid.slice(0, 37) + '...' : tid;
    console.log(`  ${tidDisplay.padEnd(40)} ${pr.toFixed(3).padStart(10)} ${ms.toFixed(3).padStart(11)} ${passAtK.toFixed(3).padStart(8)} ${passPowerK.toFixed(3).padStart(8)} ${md.toFixed(3).padStart(11)} ${mt.toFixed(0).padStart(8)} ${String(k).padStart(7)}`);
  }
}
